src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -3199,13 +3205,25 @@ void EmitLazy(RegexNode node)
                 // So, we transfer sliceStaticPos to pos, and ensure that any path out of here has sliceStaticPos as 0.
                 TransferSliceStaticPosToPos();
 
-                string startingPos = ReserveName("lazyloop_starting_pos");
-                string iterationCount = ReserveName("lazyloop_iteration");
-                string sawEmpty = ReserveName("lazyloop_empty_seen");
                 string body = ReserveName("LazyLoopBody");
                 string endLoop = ReserveName("LazyLoopEnd");
 
-                writer.WriteLine($"int {iterationCount} = 0, {startingPos} = pos, {sawEmpty} = 0;");
+                string iterationCount = ReserveName("lazyloop_iteration");
+                additionalDeclarations.Add($"int {iterationCount} = 0;");
+                writer.WriteLine($"{iterationCount} = 0;");
+
+                // Loops that match empty iterations need additional checks in place to prevent infinitely matching (since
+                // you could end up looping an infinite number of times at the same location).  We can avoid those
+                // additional checks if we can prove that the loop can never match empty, which we can do by computing
+                // the minimum length of the child; only if it's 0 might iterations be empty.
+                bool iterationMayBeEmpty = child.ComputeMinLength() == 0;
+                string startingPos = null, sawEmpty = null;
+                if (iterationMayBeEmpty)
+                {
+                    startingPos = ReserveName("lazyloop_starting_pos");
+                    sawEmpty = ReserveName("lazyloop_empty_seen");
```

Comment:
NIT: the variable name is a bit confusing since it is a `int` as opposed to a `bool`. Should we use something like `laoop_empty_seen_count` or something like that?

Reply:
It is a Boolean, in that it can only be 0 or 1.  It's just stored as an int because that's what our stack stores.